README.md
# Coronavirus Cases Scraper for Berlin

![logo for "Coronavirus Cases Scraper for Berlin" project](logo/corona_cases_berlin_small.png)

### Quick Links

* [Timeline of case numbers](data/target/berlin_corona_cases.json)
* [Timeline of traffic light indicators ("Corona Ampel")](data/target/berlin_corona_traffic_light.json)

This is a scraper for the [Corona/COVID-19 dashboard for Berlin](https://www.berlin.de/corona/lagebericht/desktop/corona.html), as issued by the [Senatsverwaltung für Gesundheit, Pflege und Gleichstellung](https://www.berlin.de/sen/gpg/) (Senate Department for Health, Care and Equality) and the [Landesamt für Gesundheit und Soziales](https://www.berlin.de/lageso/) (Regional Office for Health and Social Affairs).
The dashboard includes daily case numbers by district and age groups, as well as the "Corona traffic light"-indicators (incidence of new infections per week, ICU occupancy rate, relative change in incidence).

* Change in the 7-day incidence was dropped again on September 2nd, 2021.
Instead, the 7-day indicence for hospitalisation was introduced as a new traffic light indicator on that day.
Also, the color scheme for the incidence traffic light was adjusted: where before it was yellow at >=20% and red at >=40%, it is now yellow at >=35% and red at >=100%.
This of course makes it pointless to compare the color values before and after 2021-09-02.
* The basic reproduction number R was included until July 22nd, 2021.
After that, [R was dropped because it was no longer deemed a useful indicator](https://data.lageso.de/lageso/corona/archiv/berlin-website-2021-07-23.html "Berliner Corona-Lagebericht vom 23.07.2021").
Instead, the relative change in incidence is now the third indicator in the corona traffic light.
R is still included in the output data (as `0.0`), since some apps might expect to find it there.
* Starting with February 15, 2021, the absolute numbers and percentages for administered vaccinations are included in the dashboard.
* Starting with November 11, 2020, the change in the 7-day incidence is also included in the traffic light indicators.
* As of August 31, 2020, the dashboard [replaces the previously used daily press releases](https://www.berlin.de/sen/gpg/service/presse/2020/pressemitteilung.983232.php "SenGPG press release announcing the new Corona/COVID-19 dashboard from 2020-08-31") containing the same data.
There were two separate press releases each day, one with the case numbers (see [here](https://www.berlin.de/sen/gpg/service/presse/2020/pressemitteilung.982684.php "Last SenGPG press release with Corona/COVID-19 case numbers for Berlin from 2020-08-30") for the last one) and one with the traffic light indicators (see [here](https://www.berlin.de/sen/gpg/service/presse/2020/pressemitteilung.982682.php "Last SenGPG press release with Corona/COVID-19 traffic light indicators for Berlin from 2020-08-30") for the last one of those).


The output of the scraper are timelines of data extracted from the individual press releases.

## Output Data

_Note that there is no data for 2021-05-06, because on that day, the reporting was changed from same-day-evening to next-day-noon.
See [this note](note20210506.md) for more information._

### Corona Case Numbers

The timeline data generated by the case number scraper is a JSON file in [data/target/berlin_corona_cases.json](data/target/berlin_corona_cases.json), structured as follows:

```json
[
  {
    "date": "2020-09-23",
    "source": "https://www.berlin.de/corona/lagebericht/desktop/corona.html",
    "counts_per_district": {
      "lor_01": {
        "case_count": 2141,
        "indicence": 555.0,
        "recovered": 1894
      },
      "lor_02": {
        "case_count": 1359,
        "indicence": 468.0,
        "recovered": 1176
      },
      ...
      "lor_12": {
        "case_count": 974,
        "indicence": 365.6,
        "recovered": 897
      }
    },
    "counts_per_age_group": {
      "0-4": {
        "case_count": 358,
        "incidence": 188.7
      },
      "5-9": {
        "case_count": 406,
        "incidence": 236.5
      },
      ...
      "90+": {
        "case_count": 156,
        "incidence": 500.6
      },
      "unknown": {
        "case_count": 15,
        "indidence": "n.a."
      }
    }
  },
  {
    "date": "2020-09-22",
    "source": "https://www.berlin.de/corona/lagebericht/desktop/corona.html",
    "counts_per_district": {
      "lor_01": {
        "case_count": 2116,
        "indicence": 548.5,
        "recovered": 1877
      },
      "lor_02": {
        "case_count": 1330,
        "indicence": 458.0,
        "recovered": 1154
      },
      ...
      "lor_12": {
        "case_count": 967,
        "indicence": 363.0,
        "recovered": 890
      }
    },
    "counts_per_age_group": {
      "0-4": {
        "case_count": 356,
        "incidence": 187.6
      },
      "5-9": {
        "case_count": 399,
        "incidence": 232.4
      },
      ...
      "90+": {
        "case_count": 155,
        "incidence": 497.4
      },
      "unknown": {
        "case_count": 15,
        "indidence": "n.a."
      }
    }
  },
  ...
]
```

The structure of the data is a JSON array with objects for each day.
For each day, the `source` is specified (where was the data scraped from – this used to be a particular press release, now it is always the dashboard), the `date` (of the data), the `counts_per_district` and the `counts_per_age_group`.
For the days scraped from individual press releases, there is also a `pr_date` attribute (when was the press release issued), because `pr_date` and `date` are not always the same day.

#### Counts per District

The `counts_per_district` objects are structured with a key for each district, which in turn contain the actual numbers for the total `case_count`, `incidence` and number of `recovered` cases.

The district keys are their [LOR codes](data/manual/lor_district_codes.json) (see the dataset [Lebensweltlich orientierte Räume (LOR) in Berlin](https://daten.berlin.de/datensaetze/lebensweltlich-orientierte-räume-lor-berlin "The dataset 'Lebensweltlich orientierte Räume (LOR) in Berlin' on the Berlin Open Data Portal") for a complete definition of each LOR code):

```json
{
    "lor_01": "Mitte",
    "lor_02": "Friedrichshain-Kreuzberg",
    "lor_03": "Pankow",
    "lor_04": "Charlottenburg-Wilmersdorf",
    "lor_05": "Spandau",
    "lor_06": "Steglitz-Zehlendorf",
    "lor_07": "Tempelhof-Schöneberg",
    "lor_08": "Neukölln",
    "lor_09": "Treptow-Köpenick",
    "lor_10": "Marzahn-Hellersdorf",
    "lor_11": "Lichtenberg",
    "lor_12": "Reinickendorf"
}
```

#### Counts per Age Group

The `counts_per_age_group` objects are similarly structured, with a key for each age group, which in turn contain the `case_count` and `incidence` (no `recovered`).
The age group `80+` was split into `80-89` and `90+` beginning May 11th (2020-05-11).

There is a special `unknown` age group for which the `incidence` is always `n.a.`

#### Manually Extracted Data

Some of the earlier press releases had a slightly different format, or were only available as screen shots (true story), so the [old press release scraper](https://github.com/knudmoeller/berlin_corona_cases/tree/0.2.4 "Release 0.2.4 of the 'Coronavirus Cases Scraper for Berlin'") did not work for them.
Rather than writing special code for extracting these one- or two-off cases, I manually extracted them and put them in [data/manual/manually_extracted.json](data/manual/manually_extracted.json).
When creating the complete timeline, this manually extracted data was then merged with the newly scraped data.

### Corona Traffic Light Indicators and Vaccination

The traffic light and vaccination data generated by the scraper is a JSON file located in [data/target/berlin_corona_traffic_light.json](data/target/berlin_corona_traffic_light.json). 
This data contains a timeline of how the traffic light indicators changed over time.

Starting February 15, 2021, the Corona dashboard includes data on vaccination. This has been added to the traffic light JSON file.
While it is not strictly speaking one of the traffic light indicators, it is still an important indicator for the overall situation regarding the pandemic in Berlin.

There is a second JSON file in [data/target/berlin_corona_traffic_light.latest.json](data/target/berlin_corona_traffic_light.latest.json) which always contains the latest traffic light indicators.

The structure is as follows:

```json
[
  {
    "source": "https://www.berlin.de/corona/lagebericht/desktop/corona.html",
    "pr_date": "2021-09-02",
    "indicators": {
      "basic_reproduction_number": {
        "color": "",
        "value": 0.0
      },
      "incidence_new_infections": {
        "color": "yellow",
        "value": 83.2
      },
      "icu_occupancy_rate": {
        "color": "yellow",
        "value": 5.4
      },
      "change_incidence": {
        "color": "",
        "value": 0.0
      },
      "incidence_hospitalisation": {
        "color": "green",
        "value": 1.3
      }
    },
    "vaccination": {
      "total_administered": 4503235,
      "percentage_one_dose": 65.2,
      "percentage_two_doses": 60.4
    }
  },
  ...
  {
    "source": "https://www.berlin.de/sen/gpg/service/presse/2020/pressemitteilung.982682.php",
    "pr_date": "2020-08-30",
    "indicators": {
      "basic_reproduction_number": {
        "value": 1.1,
        "color": "green"
      },
      "incidence_new_infections": {
        "value": 12.5,
        "color": "green"
      },
      "icu_occupancy_rate": {
        "value": 1.5,
        "color": "green"
      }
    }
  },
  ...
]
```

The structure of the data is a JSON array with objects for day.
Each day specifies the `source` (where was the data scraped from – this used to be a particular press release, now it is always the dashboard), the `pr_date` (date when this particular set of indicators was announced – this used to be the date of the press release), an `indicators` object and a `vaccination` object (starting 2021-02-15).
`indicators` in turn contains the indicators `incidence_new_infections` (incidence of new infections per 100,000 inhabitants per week) and `icu_occupancy_rate` (the ICU occupancy rate in %: which percentage of the available ICU capacity is currently being used).
On 2020-11-11 another indicator was introduced: the change in 7-day incidence (_"Veränderung der 7-Tage-Inzidenz"_).
This indicator is included as `change_incidence` (the number shows the change in percent).
`change_incidence` was dropped again on 2021-09-02 (still included as `0.0` for backwards-compatibility) and replaced with a new indicator `incidence_hospitalisation`.
`basic_reproduction_number` (basic reproduction number R) was recorded until 2021-07-22.
After that, it is only included as `0.0`, in case applications rely on it to be there.

Each indicator has a numeric `value` and a traffic light `color`-code (one of [`green`, `yellow`, `red`]).
For the exact meaning of color codes please refer to the corona dashboard.
Note that the meaning of the color codes was adjusted on 2021-09-02.

`vaccination` shows the total number of administered doses of COVID-19 vaccinations, as well as the percentage of the population that has received one or two doses, respectively.

## Running the Scraper

### Running Automatically with GitHub Actions

The scraper runs automatically every day, several times starting midday (around the time when the dashboard is usually updated).
Previously, the scraper ran in the afternoon.
This changed on 2021-05-07, when the update times for the dashboard moved to noon the following day.

To run the scraper at the specified times, I have defined a workflow in [.github/workflows/scraper.yml](.github/workflows/scraper.yml) for [GitHub Actions](https://github.com/features/actions), GitHub's continuous integration framework.
The workflow 

* sets up a virtual machine,
* checks out the repository,
* runs the scraper and
* commits and pushes the updated data if there are changes.

Setting up the workflow was surprisingly easy, so I'd definitely recommend this if you want to regularly run a scraper and don't want to push the button yourself every day!

[@jaimergp](https://github.com/jaimergp) and later also [@graste](https://github.com/graste) recommended doing this, and I'm very grateful for the inspiration!
I had no idea gh-actions included a cron-based trigger that makes this possible ...

[@simonw](https://github.com/simonw)'s blog post at https://simonwillison.net/2020/Oct/9/git-scraping/ is a very good starting point if you want to learn how to do [git-scraping](https://github.com/topics/git-scraping).

### Running Manually

If you want to run the scraper yourself manually (maybe you want to improve it), you can. Here is how:

#### Requirements

* Ruby 2.7.1 (most likely works with other versions too, I haven't tried)
* the [Nokogiri](https://nokogiri.org) gem
* the [jq](https://stedolan.github.io/jq/) JSON processor (jq is not a hard requirement anymore – it's only used to extract the current day for [data/target/berlin_corona_traffic_light.latest.json](data/target/berlin_corona_traffic_light.latest.json)).

#### Installation

- First, make sure you have Ruby installed.
- If you have the [Bundler](https://bundler.io) tool, you can install the gem (Ruby library) dependencies like this:

```bash
$ bundler install
# ... output from bundler ...
```

- If you don't have bundler (you should get it), just install the dependencies individually with `gem`.
Since there is currently really only one non-standard gem dependency, this is no less convenient:

```bash
$ gem install nokogiri
# ... output from gem ...
```

- Finally, download this repository to a place of your choosing, or use `git clone`.

#### Make Targets

There is a [Makefile](Makefile) that orchestrates the scraping.
The targets should be easy to understand (each one has an `echo` statement that verbosely says what it does).

To create the complete timeline for both case numbers and traffic light indicators, use the `all` target.
You should get output like this:

```
$ make all
deleting temp folder ...
creating temp directory ...
running corona dashboard parser ...
I, [2020-09-24T22:17:27.647621 #99403]  INFO -- : reading current case number file data/target/berlin_corona_cases.json ...
I, [2020-09-24T22:17:27.664712 #99403]  INFO -- : reading current traffic light file data/target/berlin_corona_traffic_light.json ...
I, [2020-09-24T22:17:27.666040 #99403]  INFO -- : loading and parsing dashboard from https://www.berlin.de/corona/lagebericht/desktop/corona.html ...
I, [2020-09-24T22:17:30.703220 #99403]  INFO -- : getting publication date ...
I, [2020-09-24T22:17:30.712117 #99403]  INFO -- : publication date is 2020-09-24 ...
I, [2020-09-24T22:17:30.712160 #99403]  INFO -- : extracting case number data ...
I, [2020-09-24T22:17:30.733268 #99403]  INFO -- : extracting traffic light data ...
copying data from data/temp/berlin_corona_cases.json to data/target/berlin_corona_cases.json ...
copying data from data/temp/berlin_corona_traffic_light.json to data/target/berlin_corona_traffic_light.json ...
extracting latest set of traffic light indicators from data/target/berlin_corona_traffic_light.json ...
writing to data/target/berlin_corona_traffic_light.latest.json ...
write current date ...
update README.md with current date
```

## What Happened to the Old Scraper?

Initially, I had written a Python-based scraper for the daily press releases, using the [Scrapy](https://scrapy.org) Web-scraping framework.
When SenGPG stopped publishing those press releases, the scraper was no longer functional.
For a few days I "scraped" manually, until I got a new scraper for the dashboard running.
The new scraper is Ruby-based and uses [Nokogiri](https://nokogiri.org).
This doesn't mean I now think Scrapy is bad – far from it, it's great!
It's just that for the task of scraping all individual Corona press releases, starting from a paged index of press releases, Scrapy has some pretty handy functionality.
Also, I just wanted to try it out.
For simply scraping the same dashboard page every day, Scrapy seemed overkill, and I have more experience using Ruby and Nokogiri.
It was just easier for me personally.

If you're looking for the old Scrapy-based scraper, you can still find it in [release 0.2.4](https://github.com/knudmoeller/berlin_corona_cases/tree/0.2.4 "Release 0.2.4 of the 'Coronavirus Cases Scraper for Berlin'").

## Logo

- [virus](https://fontawesome.com/icons/virus) logo by [FontAwesome](https://fontawesome.com) under [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/).

## License

All software in this repository is published under the [MIT License](LICENSE).
All data in this repository (in particular the `.json` files) is published under [CC BY 3.0 DE](https://creativecommons.org/licenses/by/3.0/de/).

## Disclaimer

I do not make any claims that the data in [data/target/berlin_corona_cases.json](data/target/berlin_corona_cases.json) is correct!
If you find bugs in the code or in the data, please let me know by opening an issue [here](https://github.com/knudmoeller/berlin_corona_cases/issues).

---

2020, Knud Möller

Repository: [https://github.com/knudmoeller/berlin_corona_cases](https://github.com/knudmoeller/berlin_corona_cases)

Last changed: 2021-09-09


CHANGELOG.md
# Changelog

## Development

## 0.4.9

_(2021-09-02)_

- The `change_incidence` traffic light indicator, which was introduced in [0.4.3](CHANGELOG.md#043), is no longer included in the dashboard.
It is still included as `0.0` in the output data for backwards compatibility.
- Instead of `change_incidence`, a new indicator called `incidence_hospitalisation` is now included. 

## 0.4.8

_(2021-07-23)_

- The R-number is no longer parsed, since it is not included in the dashboard anymore (thanks [@jaimergp](https://github.com/jaimergp)).
R is still included in the output data (as `0.0`), since some apps might expect to find it there.

## 0.4.7

_(2021-05-07)_

- Adjust times for running the scraper, because the dashboard is now updated around noon, rather than in the evening. See [.github/workflows/scraper.yml](.github/workflows/scraper.yml).

## 0.4.6

_(2021-02-27)_

- Add vaccination data to the traffic light file (thanks [@jaimergp](https://github.com/jaimergp)). Vaccination data has been included since 2021-02-15.

## 0.4.5

_(2021-01-17)_

- Update Nokogiri via dependabot.

## 0.4.4

_(2020-12-17)_

- Check out a specific version of bundler to prevent the _"can't find gem bundler (>= 0.a) with executable bundle"_ error.
- Extend the cron by one hour to catch unusually late publications.

## 0.4.3

_(2020-12-06)_

- Extract change in 7-day incidence as an additional traffic light indicator.
The change in the 7-day incidence (_"Veränderung der 
7-Tage-Inzidenz"_) was introduced as an additional metric to the dashboard on 2020-11-11.

## 0.4.2

_(2020-11-24)_

- Another bugfix to adjust to differences in the markup (space as thousands separator, changed column name for incidence column).

## 0.4.1

_(2020-11-24)_

- Small bugfix (add test for `nil`  to `german_to_international_float()`) to prevent crashes due to missing values.

## 0.4.0

_(2020-10-18)_

- The scraper now runs automatically every day with GitHub Actions, [as recommended](https://github.com/knudmoeller/berlin_corona_cases/issues/1#issuecomment-698192160) by [@jaimergp](https://github.com/jaimergp).
There is a really cool little blog post on Git scraping by [@simonw](https://github.com/simonw) at https://simonwillison.net/2020/Oct/9/git-scraping/.
- Add [documentation for the GitHub Actions workflow](https://github.com/knudmoeller/berlin_corona_cases#running-automatically-with-github-actions) to the README.
- Update changelog with dates for each release.

## 0.3.3

_(2020-09-30)_

- Add the color code for a red traffic light indicator. Had to wait for an indicator to actually turn red to see what the code is. Unfortunately, this happened today (2020-09-30).

## 0.3.2

_(2020-09-24)_

- Add a Nokogiri-based (we're doing Ruby now, [because reasons](https://github.com/knudmoeller/berlin_corona_cases#what-happened-to-the-old-scraper)) scraper to extract both the case numbers and the traffic light data from the new corona dashboard.
- Update Makefile.
- Remove all Scrapy-related code.
- Update README to reflect all this.

## 0.3.1

_(2020-09-01)_

- Remove some Scrapy-related make targets.

## 0.3.0

_(2020-08-31)_

- The Senatsverwaltung stopped publishing corona press releases, so the scraper doesn't work anymore. Instead, I will now try to convert the daily JOSN with case numbers and extract the traffic light indicators from the new dashboard at https://www.berlin.de/corona/lagebericht/desktop/corona.html. The press releases weren't great, but at least there was an official record with a history of corona data. Now, there is only the data for the current day, which is lost once new data is published.
- Initially, the conversion from the new sources is done only half-automatically (it's late...).

## 0.2.4

_(2020-07-22)_

- Add quick links to data files at the top of the README).
- Add conditions to deal with bad source data for data's date (yeah).

## 0.2.3

_(2020-07-13)_

- Enable the Scrapy venv from within the Makefile. This requires a `SCRAPY_HOME` environment variable to be set (see README).

## 0.2.2

_(2020-07-12)_

- Two more case number PRs (2020-04-26 and 2020-04-11) and two traffic light PRs (2020-06-08 and 2020-05-31) had been missing because the patterns to match their titles were too restrictive. They are now included.

## 0.2.1

_(2020-07-11)_

- There were two PRs that had fallen through the cracks because they were named differently (2020-05-24 and 2020-05-25). Those two have now been added.

## 0.2.0

_(2020-06-11)_

- Add a second scraper for the Corona traffic light press releases.
- Add `pr_date` (date of press release) to case number data.
- Move some scraper helper methods up to module.
- Restructure make targets: there is now one for each scraper (`case-numbers` and `traffic-light`). Both are now triggered by `data`.

## 0.1.0

_(2020-06-07)_

- Counts per age group have been added as a new key `counts_per_age_group` for each press release that includes them (all but the very first ones).

## 0.0.3

_(2020-06-07)_

- Fix bug where case numbers with thousands separators resulted in wrong data.

## 0.0.2

_(2020-06-07)_

- Instead of extracting the press release's date for specific day of Corona case numbers, we now extract the date when the data itself was released. This was mostly identical, but in some cases not.
- Instead of extracting just a date, we now extract a datetime (no timezone).
- Logo added to project.

## 0.0.1

_(2020-06-05)_

- initial version
- contains only counts per district


coronaberlin/settings.py
# -*- coding: utf-8 -*-

# Scrapy settings for coronaberlin project
#
# For simplicity, this file contains only settings considered important or
# commonly used. You can find more settings consulting the documentation:
#
#     https://docs.scrapy.org/en/latest/topics/settings.html
#     https://docs.scrapy.org/en/latest/topics/downloader-middleware.html
#     https://docs.scrapy.org/en/latest/topics/spider-middleware.html

BOT_NAME = 'coronaberlin'

SPIDER_MODULES = ['coronaberlin.spiders']
NEWSPIDER_MODULE = 'coronaberlin.spiders'


# Crawl responsibly by identifying yourself (and your website) on the user-agent
USER_AGENT = 'coronaberlin (+https://github.com/knudmoeller/berlin_corona_cases)'

# Obey robots.txt rules
ROBOTSTXT_OBEY = True

# Configure maximum concurrent requests performed by Scrapy (default: 16)
#CONCURRENT_REQUESTS = 32

# Configure a delay for requests for the same website (default: 0)
# See https://docs.scrapy.org/en/latest/topics/settings.html#download-delay
# See also autothrottle settings and docs
#DOWNLOAD_DELAY = 3
# The download delay setting will honor only one of:
#CONCURRENT_REQUESTS_PER_DOMAIN = 16
#CONCURRENT_REQUESTS_PER_IP = 16

# Disable cookies (enabled by default)
#COOKIES_ENABLED = False

# Disable Telnet Console (enabled by default)
#TELNETCONSOLE_ENABLED = False

# Override the default request headers:
#DEFAULT_REQUEST_HEADERS = {
#   'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
#   'Accept-Language': 'en',
#}

# Enable or disable spider middlewares
# See https://docs.scrapy.org/en/latest/topics/spider-middleware.html
#SPIDER_MIDDLEWARES = {
#    'coronaberlin.middlewares.CoronaberlinSpiderMiddleware': 543,
#}

# Enable or disable downloader middlewares
# See https://docs.scrapy.org/en/latest/topics/downloader-middleware.html
#DOWNLOADER_MIDDLEWARES = {
#    'coronaberlin.middlewares.CoronaberlinDownloaderMiddleware': 543,
#}

# Enable or disable extensions
# See https://docs.scrapy.org/en/latest/topics/extensions.html
#EXTENSIONS = {
#    'scrapy.extensions.telnet.TelnetConsole': None,
#}

# Configure item pipelines
# See https://docs.scrapy.org/en/latest/topics/item-pipeline.html
#ITEM_PIPELINES = {
#    'coronaberlin.pipelines.CoronaberlinPipeline': 300,
#}

# Enable and configure the AutoThrottle extension (disabled by default)
# See https://docs.scrapy.org/en/latest/topics/autothrottle.html
#AUTOTHROTTLE_ENABLED = True
# The initial download delay
#AUTOTHROTTLE_START_DELAY = 5
# The maximum download delay to be set in case of high latencies
#AUTOTHROTTLE_MAX_DELAY = 60
# The average number of requests Scrapy should be sending in parallel to
# each remote server
#AUTOTHROTTLE_TARGET_CONCURRENCY = 1.0
# Enable showing throttling stats for every response received:
#AUTOTHROTTLE_DEBUG = False

# Enable and configure HTTP caching (disabled by default)
# See https://docs.scrapy.org/en/latest/topics/downloader-middleware.html#httpcache-middleware-settings
#HTTPCACHE_ENABLED = True
#HTTPCACHE_EXPIRATION_SECS = 0
#HTTPCACHE_DIR = 'httpcache'
#HTTPCACHE_IGNORE_HTTP_CODES = []
#HTTPCACHE_STORAGE = 'scrapy.extensions.httpcache.FilesystemCacheStorage'


coronaberlin/spiders/corona_spider.py
from collections import OrderedDict
import re
import scrapy
from coronaberlin.spiders import parse_german_float, remove_thousand_separator, extract_pr_date

class CoronaSpider(scrapy.Spider):
    name = "berlin-corona-scraper"
    start_urls = [
        "https://www.berlin.de/sen/gpg/service/presse/2020/?page_at_1_0=1",
    ]

    district_mapping = {
        "Mitte": "lor_01" ,
        "Friedrichshain-Kreuzberg": "lor_02" ,
        "Pankow": "lor_03" ,
        "Charlottenburg-Wilmersdorf": "lor_04" ,
        "Spandau": "lor_05" ,
        "Steglitz-Zehlendorf": "lor_06" ,
        "Tempelhof-Schöneberg": "lor_07" ,
        "Neukölln": "lor_08" ,
        "Neuköln": "lor_08" ,
        "Treptow-Köpenick": "lor_09" ,
        "Marzahn-Hellersdorf": "lor_10" ,
        "Lichtenberg": "lor_11" ,
        "Reinickendorf": "lor_12" ,
    }

    # regex patterns
    corona_press_release_pattern = re.compile(r'Coronavirus: Derzeit [\d\.]+ bestätigte Fälle in Berlin')
    corona_press_release_pattern2 = re.compile(r'Coronavirus in Berlin: Bestätigte Fälle und Corona-Ampel')
    datetime_pattern = re.compile(r'(\d\d)\.(\d\d)\.(\d\d\d\d), (\d\d)\:(\d\d)')
    case_count_pattern = re.compile(r'((\d|\.)+)\s*(\(.+\))?')
    german_float_pattern = re.compile(r'')
    item_pattern = re.compile(r"\d+\. ein.+({}).+".format("|".join(district_mapping.keys())))

    def parse(self, response):
        press_release_page_links = response.css('.list-autoteaser .row-fluid .text a')
        corona_case_page_links = \
            [page_link \
            for page_link \
            in press_release_page_links \
            if CoronaSpider.corona_press_release_pattern.match(page_link.css('::text').get())
             or CoronaSpider.corona_press_release_pattern2.match(page_link.css('::text').get())]

        if corona_case_page_links:
            yield from response.follow_all(corona_case_page_links, self.parse_corona_pr)

        pagination_links = response.css('li.pager-item-next a')
        if pagination_links:
            yield from response.follow_all(pagination_links, self.parse)

    def parse_corona_pr(self, response):
        result = {
            'source': response.url ,
            'pr_date': extract_pr_date(response)
        }

        # get date of data
        # first, deal with cases of bad source data:
        if response.url == "https://www.berlin.de/sen/gpg/service/presse/2020/pressemitteilung.963716.php":
            result['date'] = "2020-07-21T12:00:00"
        else:
            date_texts = response.css('.html5-section.body h2::text')
            if date_texts:
                merged = '|'.join([date_text.get() for date_text in date_texts])
                date_match = CoronaSpider.datetime_pattern.search(merged)
                iso_date = "{}-{}-{}T{}:{}:00".format(date_match.group(3), date_match.group(2), date_match.group(1), date_match.group(4), date_match.group(5))
                result['date'] = iso_date

        list_pages = {
            'https://www.berlin.de/sen/gpg/service/presse/2020/pressemitteilung.904147.php': '2020-03-08T09:30:00' ,
            'https://www.berlin.de/sen/gpg/service/presse/2020/pressemitteilung.904264.php': '2020-03-09T09:30:00'
        }

        if response.url in list_pages.keys() :
            result['date'] = list_pages[response.url]
            result = self.parse_list(response, result)
        else:
            result = self.parse_table(response, result)

        if response.url == 'https://www.berlin.de/sen/gpg/service/presse/2020/pressemitteilung.904147.php':
            result['comment'] = "Only a date has been given in this press release, no time. The time given in the 'date'-field is a guess, based on the press release from 2020-03-09."

        yield result

    def parse_table(self, response, result):
        # get first table (Bezirke)
        tables = response.css('.table-responsive')
        if not tables:
            return None
        rows = tables[0].css('tr')
        per_district = {}
        for row in rows[1:-1]:
            cells = row.css('td::text')
            district = self.parse_district(cells[0].get().strip())
            case_count = int(self.parse_case_count(cells[1].get().strip()))
            incidence = parse_german_float(cells[2].get().strip())
            per_district[district] = {
                'case_count': case_count,
                'incidence': incidence,
            }
            if len(cells) > 3:
                recovered = self.parse_recovered(cells[3].get().strip())
                per_district[district]['recovered'] = recovered

        result['counts_per_district'] = OrderedDict(sorted(per_district.items()))

        # get second table (age groups)
        if len(tables) == 2:
            rows = tables[1].css('tr')
            per_age_group = {}
            for row in rows[1:-1]:
                cells = row.css('td::text')
                age_group = cells[0].get().strip()
                case_count = int(self.parse_case_count(cells[1].get().strip()))
                incidence = "n.a."
                if age_group != "unbekannt":
                    incidence = parse_german_float(cells[2].get().strip())
                else:
                    age_group = "unknown"
                per_age_group[age_group] = {
                    'case_count': case_count,
                    'incidence': incidence,
                }

            result['counts_per_age_group'] = per_age_group

        return result

    def parse_list(self, response, result):
        per_district = {}
        for lor_key in CoronaSpider.district_mapping.values():
            per_district[lor_key] = { 'case_count': 0 }

        list_blob = response.css(".column-content .textile p:nth-of-type(3)::text")

        for item in list_blob:
            item = item.get().strip()
            item_match = CoronaSpider.item_pattern.match(item)
            if item_match:
                district = item_match.group(1)
                lor_key = CoronaSpider.district_mapping[district]
                per_district[lor_key]['case_count'] += 1

        result['counts_per_district'] = OrderedDict(sorted(per_district.items()))

        return result


    def parse_district(self, district):
        return CoronaSpider.district_mapping[district.strip()]

    def parse_case_count(self, case_count):
        count_match = CoronaSpider.case_count_pattern.match(case_count.strip())
        case_count = count_match.group(1)
        case_count = remove_thousand_separator(case_count)
        return case_count

    def parse_recovered(self, value):
        na_values = [ "n.a.", "n.a" ]
        if value in na_values:
            return "n.a."
        else:
            return int(value)
